libclc/CMakeLists.txt

```diff
@@ -116,12 +116,54 @@ execute_process( COMMAND ${LLVM_CONFIG} "--bindir"
 # These were not properly reported in early LLVM and we don't need them
 list( APPEND LLVM_CXX_FLAGS -fno-rtti -fno-exceptions )
 
+# Use custom toolchain to build libclc, this can be useful when dealing with
+# debug builds, that do not need libclc bitcode to be built using debug tools.
+if ( EXISTS ${LIBCLC_CUSTOM_LLVM_TOOLS_BINARY_DIR} )
+  message( WARNING "Using custom LLVM tools to build libclc: "
+    "${LIBCLC_CUSTOM_LLVM_TOOLS_BINARY_DIR}, "
+    " make sure that the tools are up to date." )
+
+  # find_program needs the variable to be cleared in order to perform a search.
+  # Make sure that the cached entries are cleared as well.
+  unset( LLVM_CLANG CACHE )
+  unset( LLVM_AS CACHE )
+  unset( LLVM_LINK CACHE )
+  unset( LLVM_OPT CACHE )
+  unset( LLVM_SPIRV CACHE )
+  unset( LIBCLC_REMANGLER CACHE )
+  unset( LLVM_CLANG )
+  unset( LLVM_AS )
+  unset( LLVM_LINK )
+  unset( LLVM_OPT )
+  unset( LLVM_SPIRV )
+  unset( LIBCLC_REMANGLER )
+  find_program( LLVM_CLANG clang PATHS ${LIBCLC_CUSTOM_LLVM_TOOLS_BINARY_DIR}
+    NO_DEFAULT_PATH )
+  find_program( LLVM_AS llvm-as PATHS ${LIBCLC_CUSTOM_LLVM_TOOLS_BINARY_DIR}
+    NO_DEFAULT_PATH )
+  find_program( LLVM_LINK llvm-link PATHS ${LIBCLC_CUSTOM_LLVM_TOOLS_BINARY_DIR}
+    NO_DEFAULT_PATH )
+  find_program( LLVM_OPT opt PATHS ${LIBCLC_CUSTOM_LLVM_TOOLS_BINARY_DIR}
+    NO_DEFAULT_PATH )
+  find_program( LLVM_SPIRV llvm-spirv
+    PATHS ${LIBCLC_CUSTOM_LLVM_TOOLS_BINARY_DIR} NO_DEFAULT_PATH )
+  find_program( LIBCLC_REMANGLER libclc-remangler
+    PATHS ${LIBCLC_CUSTOM_LLVM_TOOLS_BINARY_DIR} NO_DEFAULT_PATH )
+  if( NOT LLVM_CLANG OR NOT LLVM_OPT OR NOT LLVM_AS OR NOT LLVM_LINK
+      OR NOT LIBCLC_REMANGLER )
+    message( WARNING
+      "Custom toolchain incomplete, will try the default location." )
```

Comment:
This might be a case of preference, but I would prefer it to die if my path was incorrect. The warning might drown in the rest of the output and I might not realize until I get annoyed by long build times.

Reply:
The reason for not hard failing here was that if you had a vanilla llvm build (supporting required backends) you could use that for `clang` and friends, and roll back to the default path for `libclc-remangler`. But maybe that is too much of an edge case and we should fail, changed to `FATAL_ERROR` now.